src/exo/LoopIR_scheduling.py

```diff
@@ -3584,6 +3584,193 @@ def mk_write(c):
     return ir, fwd
 
 
+# TODO: maybe some feature should be added to IndexRangeEnvironment? Rather
+# than directly imporintg IndexRange and index_range_analysis
+from .range_analysis import IndexRange, index_range_analysis
+
+
+def index_range_analysis_wrapper(
+    expr: LoopIR.expr, env: IndexRangeEnvironment
+) -> IndexRange:
+    range_or_int = index_range_analysis(expr, env.env)
+    if isinstance(range_or_int, int):
+        return IndexRange.create_int(range_or_int)
+    else:
+        assert isinstance(range_or_int, IndexRange)
+        return range_or_int
+
+
+def merge_index_ranges(
+    x: Optional[IndexRange], y: Optional[IndexRange]
+) -> Optional[IndexRange]:
+    if x is None:
+        return y
+    if y is None:
+        return x
+    assert isinstance(x, IndexRange) and isinstance(y, IndexRange)
+    return x | y
+
+
+class CheckFoldBuffer(LoopIR_Do):
+    def __init__(self, proc, buffer_name, buffer_dim, size):
+        self.env = IndexRangeEnvironment(proc)
+        self.bounds = []
+        self.name = buffer_name
+        self.dim = buffer_dim
+        self.size = size
+        self.check_passed = True
+
+    def get_result(self):
+        return self.check_passed
+
+    def enter_scope(self):
+        self.env.enter_scope()
+        self.bounds.append(None)
+
+    def exit_scope(self) -> Optional[IndexRange]:
+        self.env.exit_scope()
+        return self.bounds.pop()
+
+    def validate_access(self, idx: Optional[int]) -> bool:
+        if self.bounds is None:
+            return
+        elif idx is None or self.bounds[-1].hi is None:
+            self.check_passed = False
+        else:
+            self.check_passed &= idx > self.bounds[-1].hi - self.size
+
+    def update_bounds(self, new_bounds: IndexRange, check_safety: bool = False):
+        if self.bounds[-1] is None:
+            self.bounds[-1] = new_bounds
+        else:
+            if check_safety:
+                self.validate_access(new_bounds.lo)
+            self.bounds[-1] |= new_bounds
+
+    def do_stmts(self, stmts):
+        self.enter_scope()
+        super().do_stmts(stmts)
+        return self.exit_scope()
+
+    def do_s(self, s):
+        bounds = None
+        if isinstance(s, LoopIR.For):
+            bounds = self.do_stmts(s.body)
+            env = {
+                s.iter: (
+                    s.lo.val if isinstance(s.lo, LoopIR.Const) else None,
+                    s.hi.val - 1 if isinstance(s.hi, LoopIR.Const) else None,
```

Comment:
So this doesn't handle loops like
```
for i in seq(m, m + 3):
  ...
```
?

Reply:
Fixed to support this. However, currently we cannot handle general for loops with non-constant window loop bounds.